proposals/0000-extension-lifecycle-framework.md

```diff
@@ -0,0 +1,128 @@
+
+The authors of this proposal believe that a clear lifecycle for these programming language extensions would make it easier for newcomers, those responsible for onboarding new Haskell developers, and the community more broadly, to make sense of the Haskell language extensions. This framework would make it clear which extensions are experimental and prone to rapid evolution, which of them are mature and thus relatively stable and suitable for serious use, and those that are generally not considered suitable for new development. Extending GHC's warnings to account for these categories allows projects to find out when features they use have been deprecated in a new compiler version as well as to specify general policies about language stability that can be checked mechanically.
+
+## 2. Proposed Change Specification
+
+### 2.1. Categories and Lifecycle
+Language extensions will be classified into the following categories
+ * `Experimental` extensions are undergoing active development. The syntax and semantics that are enabled by the extension are likely to change regularly. It is expected that most new language extensions will begin as experimental. At the time of writing, `LinearTypes` and `OverloadedRecordUpdate` seem to be in this category.
+ 
+ * `Mature` extensions are considered to be finished and are not expected to undergo regular changes. These features can be used without worry of unexpected changes, and they are not known to contain serious design or implementation deficiencies. Any breaking change to a mature extension will be announced well in advance of the change being made, with a migration path provided if possible. Some extensions in this category might be `MultiWayIf`, `MonoLocalBinds`, and `ViewPatterns`.
+
+ * `Deprecated` extensions are considered to be design or implementation dead ends, and should probably not be used in new code. Deprecating an extension _must_ come with a statement about the longevity of the extension. This need not be a commitment to remove it at a concrete time, but is a commitment to either remove it or transition it to another state at a time stated by GHC HQ when it is marked as `Deprecated`. Any extension will be deprecated prior to removal. Deprecated extensions are expected to include `IncoherentInstances`, `OverlappingInstances`, `Rank2Types`, `RecordPuns`, and `TypeInType`.
+ 
+ * `Legacy` extensions are explicitly not recommended for new code, much like `Deprecated`, however, they are expected to be supported indefinitely. This may be to maintain an older language set, maintain a bridge to a newer feature, or otherwise keep backward compatibility. Expected legacy extensions include `CUSKs`, `DeepSubsumption`,  and `NPlusKPatterns`.
```

Comment:
I'm struggling to see the distinction between `Deprecated` and `Legacy`. It seems the only functional difference is that they belong to different warning groups, is that right?

Given that `Deprecated` extensions do not necessarily have a concrete schedule for removal, and there is nothing stopping a `Legacy` extension subsequently being re-categorised to `Deprecated` and later removed, are they really different? I'm not convinced that trying to guess which extensions are most likely to be removed in the future is really useful. Could we simplify the proposal and have a single category for deprecated extensions?

The examples don't really help me understand: there seems to be little motivation to remove synonyms like `Rank2Types` or `RecordPuns`, whereas I can imagine changes to the typechecker making it desirable to reduce maintenance costs by removing `DeepSubsumption` at some future point.

I suppose the distinction could be made more concrete if `Deprecated` extensions were always on track for removal in some relatively predictable number of releases. But it's pretty rare for GHC to actually remove extensions in my experience (and I think this is a good thing).

Reply:
> I'm struggling to see the distinction between `Deprecated` and `Legacy`. It seems the only functional difference is that they belong to different warning groups, is that right?
> 

There is a huge difference when I've been working with people. The intent here is for users to get two different signals: 
- `Legacy`: We will continue to support this, but this is not recommended for new code. 
- `Deprecated`: This is marked for removal and is recommended to be purged from all codebases as soon as possible. If there is a compelling reason to keep it, bring that up before (timeframe set when extension was deprecated) which is the point at which it will be decided to actually remove or move to some other status.

In writing the proposal, care was taken to not be too prescriptive and concretely set a timeframe for all extensions to be removed when deprecating, as some exceptional case will certainly arise shortly after such strong rules are in place. However, we could always change the language to say the timeframe is for removal, not for the decision on removal, but that allows less community feedback on why something should be saved, no?

>  Could we simplify the proposal and have a single category for deprecated extensions?

I would strongly recommend against that. In fact I would say a huge part of the benefit of the proposal is in understanding what could actually be removed as opposed to what is not recommended but still explicitly supported going forward.

> The examples don't really help me understand: there seems to be little motivation to remove synonyms like `Rank2Types` or `RecordPuns`, whereas I can imagine changes to the typechecker making it desirable to reduce maintenance costs by removing `DeepSubsumption` at some future point.
> 
> I suppose the distinction could be made more concrete if `Deprecated` extensions were always on track for removal in some relatively predictable number of releases. But it's pretty rare for GHC to actually remove extensions in my experience (and I think this is a good thing).

Not removing extensions _ever_ actually has a large set of negative impacts. Firstly, when bringing someone to the language, we now get to say "learn Haskell" and oh learn this always increasing set of extensions that may or may not be on at any given moment when you're looking at a particular piece of code. Even if the impact to GHC is negligible for something like `Rank2Types`, the impact of having one less thing for a new user to see and trip over is actually important.  There are at least a handful of extensions that seem like good candidates to actually remove and with the warnings/notice in place from this proposal, I think that moving the forward with those deprecation and removal cycles would be beneficial to the ecosystem.

Reply:
If we have doubts that we'll ever see a extension being Deprecated, I suggest to leave it out for now. Every concept imposes a mental load on our users, company policy makers will have to make decisions (e.g. how to treat deprecated extensions). That's a cost we should not incur prematurely (or is it postMaturely?) if it's possible that we won't use it. We can still introduce the concept when it becomes relevant.

Reply:
Actually I wouldn't go so far as to say that we should *never* deprecate an extension, so I think it's helpful to retain the concept.  I just think we should do so very sparingly.

For example, I would say `IncoherentInstances` does meet the bar for deprecation and eventual removal, because it is basically always better to use per-instance pragmas. But I wouldn't remove `Rank2Types` because the cost of breaking code is greater than the cost of having an obsolete synonym for `RankNTypes` hanging around.

Reply:
I am not saying that we should _never_ deprecate, and I agree we can retain the concept – but only actually introduce it to the actual code when we actually know for certain that we are going to deprecate an extension.

_If_ we know for sure that we will deprecate, say `IncoherentInstances`, then we can introduce `-WDeprecated` right away, of course. But if we are unsure about at this point, I’d advise against introducing a category that may potentially be empty for a long while.

Reply:
Indeed I suppose it would be strange to have a warning `-WXDeprecated` that could never fire. But I guess we can resolve this when debating a concrete proposal for classification of extensions (which it sounds like will probably try to deprecate some).

Reply:
Even an unused `Deprecated` category has a purpose: it communicates that all these `Legacy` extensions (those marked "deprecated" in the manual today) are not in fact scheduled for removal, and that your code that uses them is not in an emergency state. Today, "deprecated" in software development (and much other Haskell practice) usually means "removal is on the way", but if users can rely on the alias `RecordPuns` sticking around essentially forever, then it's good for them to know that.

Thinking more about this, perhaps we should also write a quick sentence for each `Legacy` or `Deprecated` extension when doing the classification proposal that can be included in the warning. An extension might be considered to be a bad idea for technical reasons (`IncoherentInstances`), or just because it's been renamed to something else (`RecordPuns`). It would be nice if the warning could say `Rank2Types` is a synonym for `RankNTypes` in GHC and that `IncoherentInstances` should probably be replaced by pragmas.

Reply:
Yes, absolutely: no proposal should ever be understood as forbidding better error messages! :-)

Maybe we should wait with “Do we actually need Deprecated or not” until after we have categorized all existing extensions (a daunting task anyways, worse than defining GHC20xx…).

Reply:
It sounds to me like there may be an actual plan to remove `TypeInType` - it's newly deprecated in 9.6.1. But we really do need to have a deliberate discussion about this in the follow-up.

Reply:
In any case, I'm marking this thread Resolved, as Adam's concrete formulations are in the current draft, and the authors will keep the `Deprecated` category.